src/goesproc/gradient.h

```diff
@@ -0,0 +1,170 @@
+  float hsv[3];
+
+  GradientPoint fromRGB(float r, float g, float b) const {
+    float min, max;
+    GradientPoint out;
+    out.units = units;
+
+    // Rescale if max input range exceeds 1
+    if (r > 1 || g > 1 || b > 1) {
+      r /= 255;
+      g /= 255;
+      b /= 255;
+    }
+
+    out.rgb[0] = r;
+    out.rgb[1] = g;
+    out.rgb[2] = b;
+
+    // Compute min/max RGB
+    if (r <= g && r <= b) min = r;
+    else if (g <= r && g <= b) min = g;
+    else min = b;
+
+    if (r >= g && r >= b) max = r;
+    else if (g >= r && g >= b) max = g;
+    else max = b;
+
+    // Perform HSV->RGB conversion
+    out.hsv[2] = max;
+    float d = max - min;
+    out.hsv[1] = (max <= 0) ? 0 : d / max;
+
+    if (min == max) out.hsv[0] = 0;
+    else {
+      if (max == r) out.hsv[0] = fmod((g - b) / d, 6);
+      else if (max == g) out.hsv[0] = (b - r) / d + 2;
+      else if (max == b) out.hsv[0] = (r - g) / d + 4;
+    }
+
+    out.hsv[0] /= 6;
+    return out;
+  }
+
+  GradientPoint fromHSV(float h, float s, float v) const {
+    GradientPoint out;
+    out.units = units;
+
+    // Rescale if max input range exceeds 1
+    if (h > 1 || s > 1 || v > 1) {
+      h /= 360;
+      s /= 100;
+      v /= 100;
+    }
+
+    out.hsv[0] = h;
+    out.hsv[1] = s;
+    out.hsv[2] = v;
+
+    // Perform RGB->HSV conversion
+    int i = (int) (h * 6);
+    float f = h * 6 - i;
+    float p = v * (1 - s);
+    float q = v * (1 - f * s);
+    float t = v * (1 - (1 - f) * s);
+
+    switch (i % 6) {
+      case 0: out.rgb[0] = v; out.rgb[1] = t; out.rgb[2] = p; break;
+      case 1: out.rgb[0] = q; out.rgb[1] = v; out.rgb[2] = p; break;
+      case 2: out.rgb[0] = p; out.rgb[1] = v; out.rgb[2] = t; break;
+      case 3: out.rgb[0] = p; out.rgb[1] = q; out.rgb[2] = v; break;
+      case 4: out.rgb[0] = t; out.rgb[1] = p; out.rgb[2] = v; break;
+      case 5: out.rgb[0] = v; out.rgb[1] = p; out.rgb[2] = q; break;
+    }
+
+    return out;
+  }
+};
+
+std::ostream& operator<<(std::ostream& os, const GradientPoint &p);
+
+enum GradientInterpolationType {
+  LERP_UNDEFINED = -1,
+  LERP_RGB,
+  LERP_HSV,
+  LERP_NUMTYPES
+};
+
+struct Gradient {
+  std::vector<GradientPoint> points;
+
+  inline int size() const {
+    return points.size();
+  }
+
+  // Search for the right place to insert a new point
+  int addPoint(const GradientPoint &point) {
+    auto i = points.begin();
+    while (i != points.end() && i->units < point.units) i++;
+    points.insert(i, point);
+    return size();
+  }
+
+  // Emit gradient points to stderr for debug purposes
+  void debug() const {
+    for (auto i = points.begin(); i != points.end(); i++) {
+      std::cerr << "[" << (i - points.begin()) << "] -> " << *i << std::endl;
+    }
+  }
+
+  // Find bounding points and perform linear interpolation
+  GradientPoint interpolate(const float units, const GradientInterpolationType lerptype = LERP_UNDEFINED) const {
+    auto left = points.end();
+    auto right = points.begin();
+
```

Comment:
Because the default mode is LERP_RGB, can you add an if lerpType == LERP_UNDEFINED -> lerpType = LERP_RGB here? Then it's not the else branch below and both types are covered.

Reply:
Perhaps adding the LERP_UNDEFINED case to the if()?